frame/treasury/src/lib.rs

```diff
@@ -457,6 +826,146 @@ mod tests {
+
+			assert_ok!(Treasury::tip(Origin::signed(12), h.clone(), 10));
+			assert_noop!(Treasury::close_tip(Origin::signed(0), h.into()), Error::<Test>::Premature);
+
+			System::set_block_number(2);
+			assert_noop!(Treasury::close_tip(Origin::NONE, h.into()), BadOrigin);
+			assert_ok!(Treasury::close_tip(Origin::signed(0), h.into()));
+			assert_eq!(Balances::free_balance(&3), 10);
+
+			assert_noop!(Treasury::close_tip(Origin::signed(100), h.into()), Error::<Test>::UnknownTip);
+		});
+	}
+
+	#[test]
+	fn retract_tip_works() {
+		new_test_ext().execute_with(|| {
+			// Check that accumulate works when we have Some value in Dummy already.
+			Balances::make_free_balance_be(&Treasury::account_id(), 101);
+			assert_ok!(Treasury::report_awesome(Origin::signed(0), b"awesome.dot".to_vec(), 3));
+			let h = tip_hash();
+			assert_ok!(Treasury::tip(Origin::signed(10), h.clone(), 10));
+			assert_ok!(Treasury::tip(Origin::signed(11), h.clone(), 10));
+			assert_ok!(Treasury::tip(Origin::signed(12), h.clone(), 10));
+			assert_noop!(Treasury::retract_tip(Origin::signed(10), h.clone()), Error::<Test>::NotFinder);
+			assert_ok!(Treasury::retract_tip(Origin::signed(0), h.clone()));
+			System::set_block_number(2);
+			assert_noop!(Treasury::close_tip(Origin::signed(0), h.into()), Error::<Test>::UnknownTip);
+		});
+	}
+
+	#[test]
+	fn tip_median_calculation_works() {
+		new_test_ext().execute_with(|| {
+			// Check that accumulate works when we have Some value in Dummy already.
+			Balances::make_free_balance_be(&Treasury::account_id(), 101);
+			assert_ok!(Treasury::tip_new(Origin::signed(10), b"awesome.dot".to_vec(), 3, 0));
+			let h = tip_hash();
+			assert_ok!(Treasury::tip(Origin::signed(11), h.clone(), 10));
+			assert_ok!(Treasury::tip(Origin::signed(12), h.clone(), 1000000));
+			System::set_block_number(2);
+			assert_ok!(Treasury::close_tip(Origin::signed(0), h.into()));
+			assert_eq!(Balances::free_balance(&3), 10);
+		});
+	}
+
+	#[test]
+	fn tip_changing_works() {
+		new_test_ext().execute_with(|| {
+			// Check that accumulate works when we have Some value in Dummy already.
+			Balances::make_free_balance_be(&Treasury::account_id(), 101);
+			assert_ok!(Treasury::tip_new(Origin::signed(10), b"awesome.dot".to_vec(), 3, 10000));
+			let h = tip_hash();
+			assert_ok!(Treasury::tip(Origin::signed(11), h.clone(), 10000));
+			assert_ok!(Treasury::tip(Origin::signed(12), h.clone(), 10000));
+			assert_ok!(Treasury::tip(Origin::signed(13), h.clone(), 0));
+			assert_ok!(Treasury::tip(Origin::signed(14), h.clone(), 0));
+			assert_ok!(Treasury::tip(Origin::signed(12), h.clone(), 1000));
+			assert_ok!(Treasury::tip(Origin::signed(11), h.clone(), 100));
+			assert_ok!(Treasury::tip(Origin::signed(10), h.clone(), 10));
+			System::set_block_number(2);
+			assert_ok!(Treasury::close_tip(Origin::signed(0), h.into()));
+			assert_eq!(Balances::free_balance(&3), 10);
```

Comment:
To me this does not really check that tip changing does work.

Either I would want to look at the tip values in storage to see the value updated, or I would want to see a final resulting reward that would have changed given the tips changing.

In this case, if tips do not update `(0, 0, 10, 10000, 10000)` or  tips do update `(0, 0, 10, 100, 1000)`, the result of this test would be the same.

Reply:
if tips did not update, then the tips would be 10000,10000,10000,0,0 and the result tip would be 10.

Reply:
Of course, I was being stupid and missed the original tip from `tip_new`:

```
assert_ok!(Treasury::tip_new(Origin::signed(10), b"awesome.dot".to_vec(), 3, 10000));